base/array.jl

```diff
@@ -2329,19 +2329,39 @@ function findall(A)
 end
 
 # Allocating result upfront is faster (possible only when collection can be iterated twice)
-function findall(A::AbstractArray{Bool})
-    n = count(A)
+function _findall(f::Function, A::AbstractArray{Bool})
+    n = count(f, A)
     I = Vector{eltype(keys(A))}(undef, n)
+    _findall(f, I, A)
+end
+
+function _findall(f::Function, I::Vector, A::AbstractArray{Bool})
     cnt = 1
-    for (i,a) in pairs(A)
-        if a
-            I[cnt] = i
-            cnt += 1
-        end
+    len = length(I)
+    for (k, v) in pairs(A)
+        cnt > len && break
+        I[cnt] = k
+        cnt += f(v)
+    end
+    I
+end
+
+function _findall(f::Function, I::Vector, A::AbstractVector{Bool})
+    i = firstindex(A)
+    cnt = 1
+    len = length(I)
+    @inbounds while cnt ≤ len
+        I[cnt] = i
```

Comment:
This line can be `@inbounds` but the next 2 can't. `A` can be any AbstractArray, so can have broken indexing.

Reply:
We discussed this above. AFAIK the policy is that we use `@inbounds` when it only relies on the `AbstractArray` implementation being correct. Otherwise fallback definitions couldn't be fast. It would be too bad to force each array package to redefine all these methods just to add `@inbounds`.

Reply:
I don't know if that issue is "officially" settled. This issue is relevant. https://github.com/JuliaLang/julia/issues/42030. FWIW, I think having default implementation not be able to assume that the interface is implemented correctly is way too restrictive. It's also inconsistent: It's easy to wreck Julia by writing wrong implementations anyway.

Reply:
There are many expressions in this loop, and while there is a good argument that `A[i]` and `I[cnt]` are always inbounds (since they are checked in the loop iteration), the other expressions are not inbounds.

Reply:
Okay, I'm moving the `@inbounds` annotation to only the `Vector` since that is surely guaranteed.